Support PDU Session Establishment request received from AMF#7215

prabinakpattnaik opened the pull request:
Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

## Summary

Support PDU session establishment functionality from magma-5G to master branch.
Below functionalities are supported:
a)	INITIAL_REQUEST: Need to create the session context in SessionMap and write to memory by SessionStore.
         Initialization of SessionState for IMSI for 1st Requested session from UE with IMSI.
         Writing of SessionMap in memory through SessionStore object
b)	Update the state if session state is in "CREATING" state:
Get the UE IP address, Gnb teid value then update the sessionstore.
Make complete PDR and send the session creation request to UPF.
c)	Handle the response notification from upf w.r.t. UPF message.
d)	Update the state change notification to AMF
      Send message to AMF gRPC client handler.

## Test Plan
Below stub is used for functionality verification
lte/gateway/python/scripts/smf-upf_intergartion_TC_Stub.py

For UT framwork:
lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp.
[FT_for_PDU_Establishment_logs.txt](https://github.com/magma/magma/files/6580401/FT_for_PDU_Establishment_logs.txt)

themarwhal commented on lte/gateway/c/session_manager/sessiond_main.cpp:
```diff
@@ -205,11 +205,15 @@ int main(int argc, char* argv[]) {
       (config["converged_access"].as<bool>())) {
     converged_access = true;
   }
+  if (config["session_rtx_count"].IsDefined()) {
```
What if it's not defined? Do we have a default value?

prabinakpattnaik replied:
Yes, Session manager max resending throttle count when the UPF session info version is not matching with UPF received version no.
session_rtx_count: 3

themarwhal replied:
Not sure if I understand. Do we default to a value if the config is not specified? Is 3 the default value?

prabinakpattnaik replied:
Yes, The value was added in lte/gateway/configs/sessiond.yml.

themarwhal replied:
I think it would be good to have a default value in the code as well. Or at least a warning message to indicate the value is missing.

prabinakpattnaik replied:
Ok. I will add warning message and default value as well

prabinakpattnaik replied:
Fixed it.

themarwhal commented on lte/gateway/c/session_manager/sessiond_main.cpp:
```diff
@@ -205,11 +205,15 @@ int main(int argc, char* argv[]) {
   initialize_sentry();
 
   bool converged_access = false;
+  extern uint32_t session_max_rtx_count;
```
Instead of defining a global variable, can we pass it into the relevant object/class?

prabinakpattnaik replied:
Yes, Removed the external variable and pass in a object.

themarwhal commented on lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:
```diff
@@ -11,52 +11,68 @@
     mconfig.set_log_level(magma::orc8r::LogLevel::INFO);
     mconfig.set_relay_enabled(false);
 
+    auto config =
```
I think we should mock out this config loading step. Can we manually construct the config value we want here instead of loading it from the machine?

prabinakpattnaik replied:
Yes. defined the value here instead of fetching from config.

themarwhal commented on lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:
```diff
@@ -11,52 +11,68 @@
 
+    auto config =
+        magma::ServiceConfigLoader{}.load_service_config(SESSIOND_SERVICE);
+    magma::service303::MagmaService server(SESSIOND_SERVICE, SESSIOND_VERSION);
```
How is this service used?

prabinakpattnaik replied:
No Service used. By mistake i kept here. 
Removed it.

themarwhal commented on lte/protos/session_manager.proto:
```diff
@@ -90,14 +90,31 @@ message RatSpecificContext {
   }
 }
 
+message TeidSet {
```
Let's place any comments in the line above 
```
// Obsoelte to be cleared once AMF code clean up happens
uint32 teid = 1;
```
See [here](https://docs.magmacore.org/docs/next/contributing/contribute_conventions#grpc) for grpc/proto style guidelines. 

(Also obsolete spelling)

prabinakpattnaik replied:
Yes. I corrected all.

themarwhal commented on lte/gateway/c/session_manager/SessionStateEnforcer.h:
```diff
@@ -105,17 +137,27 @@ class SessionStateEnforcer {
+  // update the GNB endpoint details in a rule
+  bool insert_pdr_from_core(
+      const std::string& imsi, std::unique_ptr<SessionState>& session_state,
+      SessionStateUpdateCriteria& session_uc, SetGroupPDR& rule);
```
Usually we tend to place any update criteria related fields at the very end.
```
bool insert_pdr_from_core(
      const std::string& imsi, std::unique_ptr<SessionState>& session_state,
      SetGroupPDR& rule, SessionStateUpdateCriteria& session_uc)
```

prabinakpattnaik replied:
Yes, modified it.

themarwhal commented on lte/gateway/python/scripts/smf-upf_intergartion_TC_Stub.py:
```diff
@@ -0,0 +1,385 @@
+
+"""
+Copyright 2020 The Magma Authors.
+
```
Please mind the lint warnings in this file
See our contributing guide: https://docs.magmacore.org/docs/next/contributing/contribute_conventions#python

To run the linter locally,
```
# install docker
cd magma/lte/gateway/python
./precommit.py --build
./precommit.py --lint -p lte/gateway/python/scripts/smf-upf_intergartion_TC_Stub.py
./precommit.py --format -p lte/gateway/python/scripts/smf-upf_intergartion_TC_Stub.py
```

prabinakpattnaik replied:
I corrected the format of python file and removed most of lint warning.

themarwhal replied:
looks beautiful! thanks :)

themarwhal reviewed (changes requested):
Posted some initial comments. Please mind the linter annotations :)

prabinakpattnaik commented:
> Posted some initial comments. Please mind the linter annotations :)

Thank you so much Marie for provided the quick feedback.

themarwhal commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -235,26 +235,64 @@ uint32_t SessionState::get_current_version() {
 }
 
 void SessionState::set_current_version(
-    int new_session_version, SessionStateUpdateCriteria* session_uc) {
```
Let's keep the `SessionStateUpdateCriteria* session_uc` pointer type here (output parameter that can be optional should always be pointers so it's clear from the signature)

prabinakpattnaik replied:
Yes, changed as a pointer type.

themarwhal commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -235,26 +235,64 @@ uint32_t SessionState::get_current_version() {
-void SessionState::insert_pdr(SetGroupPDR* rule) {
-  PdrList_.push_back(*rule);
+void SessionState::insert_pdr(
+    SetGroupPDR* rule, SessionStateUpdateCriteria& session_uc, bool crit_add) {
```
let's change the type here to
```
void SessionState::insert_pdr(
	SetGroupPDR* rule, bool crit_add, SessionStateUpdateCriteria* session_uc)
```

prabinakpattnaik replied:
Changes done for all methods.

themarwhal commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -235,26 +235,64 @@ uint32_t SessionState::get_current_version() {
+  }
+}
+
+int32_t SessionState::pdr_index(uint32_t id) {
```
instead of returning -1 on failure, we can return `optional<uint32_t>`

prabinakpattnaik replied:
It's getting compilation because index number used in a vector. If possible, I will do fix later.

themarwhal commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -438,6 +483,20 @@ bool SessionState::apply_update_criteria(
+    Pdr_index = pdr_index(rule.pdr_id());
+    if (Pdr_index != -1) {
+      // Update the existing value
+      PdrList_.at(Pdr_index) = rule;
```
this seems like a map interface? Why don't we change the type of `PdrList_` to a map?

prabinakpattnaik replied:
PdrList_ as a vector of SetGroupPDR

themarwhal commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -2652,6 +2729,20 @@ void SessionState::clear_session_metrics() const {
+ * session version no, then we continue to resend till SESSION_THROTTLE_CNT
+ * reaches
+ */
+uint32_t SessionState::get_rtx_counter() {
```
`get_incremented_rtx_counter`

prabinakpattnaik replied:
changes done.

themarwhal commented on lte/gateway/c/session_manager/SessionState.h:
```diff
@@ -155,19 +155,32 @@ class SessionState {
 
   void set_current_version(
-      int new_session_version, SessionStateUpdateCriteria* session_uc);
+      int new_session_version, SessionStateUpdateCriteria& session_uc);
```
Let's keep the pointer type here

prabinakpattnaik replied:
Yes, Changes done

themarwhal commented on lte/gateway/c/session_manager/SessionState.h:
```diff
@@ -155,19 +155,32 @@ class SessionState {
 
-  std::vector<SetGroupPDR>& get_all_pdr_rules();
+  /* method to remove all pdrs */
+  void remove_all_rules(SessionStateUpdateCriteria& session_uc);
```
pointer parameter

prabinakpattnaik replied:
Yes. Fixed it

themarwhal commented on lte/gateway/c/session_manager/SessionState.h:
```diff
@@ -466,7 +483,7 @@ class SessionState {
       const std::string& rule_id, const RuleLifetime& lifetime);
 
-  SessionFsmState get_state() { return curr_state_; }
+  SessionFsmState get_state();
```
should be a const function

prabinakpattnaik replied:
Fixed it.

themarwhal commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -235,26 +235,64 @@ uint32_t SessionState::get_current_version() {
+    SetGroupPDR* rule, SessionStateUpdateCriteria& session_uc, bool crit_add) {
+  // Check if it already exists
+  int32_t Pdr_index;
+  Pdr_index = pdr_index(rule->pdr_id());
```
is `pdr_index` a constructor? if it's a function let's rename it to `get_pdr_index`

prabinakpattnaik replied:
Yes. Fixed it. Sorry for wrong naming conversion.

themarwhal replied:
no worries :) Thanks for fixing.

themarwhal commented on lte/gateway/c/session_manager/SessionStateEnforcer.cpp:
```diff
@@ -101,62 +115,128 @@ bool SessionStateEnforcer::m5g_init_session_credit(
+  SessionStateUpdateCriteria session_uc = session_update[imsi][session_id];
+  gnode_teid                            = update_session_rules(
+      imsi, session_state, session_uc, gnb_teid_get, upf_teid_get);
+  if (!gnode_teid) return false;
```
here and everywhere else, but I tend to prefer
```
if (!gnode_teid) {
   return false;
}
```

prabinakpattnaik replied:
Yes, Fixed it.

themarwhal commented on lte/gateway/c/session_manager/SessionStateEnforcer.cpp:
```diff
@@ -101,62 +115,128 @@ bool SessionStateEnforcer::m5g_init_session_credit(
-  auto itp        = pdr_map_.equal_range(imsi);
+    std::unique_ptr<SessionState>& session_state,
+    SessionUpdate& session_update) {
+  uint32_t gnode_teid = 0;
```
do we have to initialize `gnode_teid` here?

prabinakpattnaik replied:
Not required here.

themarwhal commented on lte/gateway/c/session_manager/SessionStateEnforcer.cpp:
```diff
@@ -101,62 +115,128 @@ bool SessionStateEnforcer::m5g_init_session_credit(
-    if (!l_teid) l_teid = pipelined_client_->get_current_teid();
-    rule.mutable_pdi()->set_local_f_teid(pipelined_client_->get_next_teid());
+    // Get the UE ip address
+    rule.mutable_pdi()->set_ue_ip_adr(
```
Let's pull the `rule.mutable_*` logic to construct the object into a separate function to clean things up.

prabinakpattnaik replied:
Created separate function for same.

themarwhal commented on lte/gateway/c/session_manager/SessionStateEnforcer.cpp:
```diff
@@ -101,62 +115,128 @@ bool SessionStateEnforcer::m5g_init_session_credit(
+  /* Irrespective of any State of Session, release and terminate*/
+  SessionUpdate session_update =
+      SessionStore::get_default_session_update(session_map);
+  SessionStateUpdateCriteria session_uc = session_update[imsi][session_id];
```
I think `SessionStateUpdateCriteria session_uc = session_update[imsi][session_id];` might be a no-op here. You either want a reference to the object in the map or a pointer. (I prefer pointer for extra clarity.)

themarwhal replied:
I'm a bit lost in the logic here but do we even need to fill in `SessionUpdate` here? If not let's pass in `nullptr` into `update_session_rules` so that it's crystal clear it's a no-op.

prabinakpattnaik replied:
Modified as a pointer type.

themarwhal reviewed (changes requested):
Thanks for addressing some comments :) 
Few more general feedback throughout the PR are:
1. move `session_uc` output parameters to be the last parameter (This is how it is done for most functions in SessionD. output parameters tend to be the last few args.
2. all mutable output parameters should be pointer types. Especially if they are optional sometimes. So this means all `SessionStateUpdateCriteria& session_uc` should be `SessionStateUpdateCriteria* session_uc`

prabinakpattnaik commented:
> Thanks for addressing some comments :)
> Few more general feedback throughout the PR are:
> 
> 1. move `session_uc` output parameters to be the last parameter (This is how it is done for most functions in SessionD. output parameters tend to be the last few args.
> 2. all mutable output parameters should be pointer types. Especially if they are optional sometimes. So this means all `SessionStateUpdateCriteria& session_uc` should be `SessionStateUpdateCriteria* session_uc`

Yes, Marie . Addressed review comments. Thanks for quick response.

themarwhal commented:
> > Thanks for addressing some comments :)
> > Few more general feedback throughout the PR are:
> > 
> > 1. move `session_uc` output parameters to be the last parameter (This is how it is done for most functions in SessionD. output parameters tend to be the last few args.
> > 2. all mutable output parameters should be pointer types. Especially if they are optional sometimes. So this means all `SessionStateUpdateCriteria& session_uc` should be `SessionStateUpdateCriteria* session_uc`
> 
> Yes, Marie . Addressed review comments. Thanks for quick response.

Thanks! Will take another go around later today.

themarwhal commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -240,21 +240,61 @@ void SessionState::set_current_version(
+
+/* method to change the PDR state */
+void SessionState::set_all_pdrs(enum PdrState pdr_state) {
+  for (auto& rule : PdrList_) {
```
`PdrList_` -> `pdr_list_` 
I also think this should be `std::unordered_map<std::string, SetGroupPDR>` if we're doing these kinds of search operations. What do you think? If the change is too destructive, we can file an issue and tackle it in a separate PR.

prabinakpattnaik replied:
PdrList_ is vector of SetGroupPDR and single IMSI could be more than one SetGroupPDRs (For uplink and downlink).
We can use unordered_multimap. I am thinking first all basic PRs should be merged into master then we will do change based on performance and best approaches of data structure logic.

themarwhal commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -925,17 +986,35 @@ TotalCreditUsage SessionState::get_total_credit_usage() {
 
-uint32_t SessionState::get_local_teid() const {
-  return local_teid_;
+std::string SessionState::get_subscriber_id() const {
```
Is this not the same as `get_imsi` ? https://sourcegraph.com/github.com/magma/magma/-/blob/lte/gateway/c/session_manager/SessionState.h#L253

prabinakpattnaik replied:
Yes. Same as get_imsi(). I removed the get_subscriber_id() and used the get_imsi().

themarwhal commented on lte/gateway/c/session_manager/SessionState.h:
```diff
@@ -254,10 +267,14 @@ class SessionState {
+  uint32_t get_pdu_id() const;
 
-  void set_local_teid(uint32_t teid, SessionStateUpdateCriteria* session_uc);
+  uint32_t get_upf_local_teid() const;
```
We should have one line between each function declarations.

prabinakpattnaik replied:
Yes, changes done.

themarwhal commented on lte/protos/session_manager.proto:
```diff
@@ -792,33 +816,54 @@ enum SMSessionFSMState {
+  // TS-29-502 - 6.1.6.3.10 
+  DnnSelectionMode dnn_selection_mode = 12;
+  // TS-29.571 5.6.
+  string trace_requirment = 13;
```
"requirment" is a misspelling of "requirement"

prabinakpattnaik replied:
Fixed it.

themarwhal commented on lte/protos/session_manager.proto:
```diff
@@ -792,33 +816,54 @@ enum SMSessionFSMState {
-  RequestType rquest_type = 2;        //Define ENUM 24-501 - 9.11.3.47 . 1 byte. i.e. initial requerst, existing pdusession etc...
-  RedirectServer pdu_address = 3; //IP address. Imported from session_manager.proto file
+  uint32 pdu_session_id = 1;
+ //Define ENUM 24-501 - 9.11.3.47 . 1 byte. i.e. initial requerst, existing pdusession etc...
```
All comments should have a space after the `//`
```suggestion
 // Define ENUM 24-501 - 9.11.3.47 . 1 byte. i.e. initial requerst, existing pdusession etc...
```

prabinakpattnaik replied:
Changes done for both proto files.

themarwhal reviewed (changes requested):
I think this PR is mostly good to go. 
1. I left some comments about pdr_list data structure. It looks to me, from the way it's being used, that it should be a map type rather than a vector. But if this is too big of a change, I think we can commit this and create a separate PR to fix it.
2. Some nit about adding a space after `//` in .proto comments.
3. One spelling thing in proto

prabinakpattnaik commented:
> I think this PR is mostly good to go.
> 
> 1. I left some comments about pdr_list data structure. It looks to me, from the way it's being used, that it should be a map type rather than a vector. But if this is too big of a change, I think we can commit this and create a separate PR to fix it.
> 2. Some nit about adding a space after `//` in .proto comments.
> 3. One spelling thing in proto

First comment will address on separate PR after merging all basic PRs. 
Thank you Marie.

themarwhal commented:
(passed) Re-running the CWAG job: https://jenkins.magmacore.org/job/CWAG-integration-test-libvirt/3644/

themarwhal commented:
Hey @prabinakpattnaik can you resolve the merge conflict? 
And @ardzoht could you take a pass at this?

prabinakpattnaik commented:
> Hey @prabinakpattnaik can you resolve the merge conflict?
> And @ardzoht could you take a pass at this?

Ok Marie, I will do resolve.

ulaskozat commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -133,14 +132,15 @@ SessionState::SessionState(
       pdp_end_time_(marshaled.pdp_end_time),
       // 5G session version handlimg
       current_version_(marshaled.current_version),
+      // SMF-UPF version missmatch, max re-sending throttle count
```
typo `missmatch`

ulaskozat replied:
max re-sending throttle count -> retransmission counter ?

prabinakpattnaik replied:
Yes, it is retransmission counter. Corrected it.

ulaskozat commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -240,21 +240,61 @@ void SessionState::set_current_version(
   if (session_uc) {
     session_uc->is_current_version_updated = true;
     session_uc->updated_current_version    = new_session_version;
+    MLOG(MINFO) << " Current version is " << get_current_version();
```
demote to MDEBUG?

prabinakpattnaik replied:
Yes, missed the debug level change. Now fixed it.

ulaskozat commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -290,12 +335,18 @@ magma::lte::Fsm_state_FsmState SessionState::get_proto_fsm_state() {
   info->ver_no              = get_current_version();
+  info->local_f_teid        = get_upf_local_teid();
   info->nodeId.node_id_type = SessionInfo::IPv4;
+  info->Pdr_rules_          = get_all_pdr_rules();
```
Pdr_rules_ -> pdr_rules?

prabinakpattnaik replied:
change done.

ulaskozat commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -438,6 +485,20 @@ bool SessionState::apply_update_criteria(
 
+  // Converged 5G rules to install
+  for (const auto& rule : session_uc.pdrs_to_install) {
+    int32_t Pdr_index;
```
Pdr_index -> pdr_index

prabinakpattnaik replied:
Yes, changed it.

ulaskozat commented on lte/gateway/c/session_manager/SessionState.h:
```diff
@@ -719,6 +737,8 @@ class SessionState {
+  uint32_t rtx_counter_;
   // All 5G specific rules
   // use as shared_ptr to check
   std::vector<SetGroupPDR> PdrList_;
```
pdr_list_

prabinakpattnaik replied:
Changed it.

ulaskozat commented on lte/protos/session_manager.proto:
```diff
@@ -789,55 +813,76 @@ enum SMSessionFSMState {
                 rpc SetAmfSessionContext (SetSMSessionContext) returns (SmContextVoid);
         }
 
 ////////// SET MESSAGE PROTO DEFINATION OF SMF TO AMF ON RESPONSE /////////////
```
typo DEFINATION. Also all these ////... look too much....

prabinakpattnaik replied:
Fixed it.

ulaskozat commented on lte/protos/session_manager.proto:
```diff
@@ -851,28 +896,48 @@ message RatSpecificContextAccess {
+   Kbps_4 = 2;
+   Kbps_16 = 3;
+   Kbps_64 = 4;
+}
```
move this inside the message definition

prabinakpattnaik replied:
Yes, moved into message body.

prabinakpattnaik commented:
Rebased with latest master.

ulaskozat reviewed (approved):
LGTM

prabinakpattnaik committed:
Support PDU Session Establishment request received from AMF#7215

Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

prabinakpattnaik commented:
> LGTM

Rebased with latest master and fixed the all conflicts.

ardzoht commented on lte/protos/session_manager.proto:
```diff
@@ -790,55 +814,76 @@ enum SMSessionFSMState {
+  AccessType access_type = 5;
+  // pei - permanent equipment ID. Variable length of 16 bytes. TS-23.003 - 6.4 Like IMEI or IMEISV
+  string imei = 6;
+  // Lenght is 10 bytes + 2 bytes country code generic public subscription id.
```
legit typo on Length

prabinakpattnaik replied:
Sorry for typo. Fixed it.

ardzoht commented on lte/gateway/c/session_manager/SessionState.cpp:
```diff
@@ -238,27 +238,72 @@ void SessionState::set_current_version(
+}
+
+/* method to search specific pdr id existence */
+bool SessionState::search_pdr(unsigned int id) {
```
`contains_pdr` seems better naming for this function

prabinakpattnaik replied:
Yes. changes done.

ardzoht reviewed (approved):
left some minor nits, please take a look before merging, overall LGTM

prabinakpattnaik committed:
Support PDU Session Establishment request received from AMF#7215

Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>
